### ISSUE TITLE
New clients are added to the seleceted tag instead of the one they were spawned from.

### ISSUE THREAD

**Issue by zhrts** [awesomeWM/awesome#1457]
Output of `awesome --version`:
```
awesome v4.0 (Harder, Better, Faster, Stronger)
 • Compiled against Lua 5.3.3 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • RandR 1.5 support: ✔
 • LGI version: 0.9.1
```

**How to reproduce the issue:**
Run a slower program like firefox and then switch tags before the window is created.

**Actual result:**
The program spawns on the currently selected tag.

**Expected result:**
The program spawns on the tag it was spawned from.

my "fix" that is probably unreliable:
```
function ()
                  awful.prompt.run {
                      prompt       = "Open: ",
                      textbox      = awful.screen.focused().mypromptbox.widget,
                      exe_callback = function(...)
                        awful.screen.focused().mypromptbox:spawn_and_handle_error(..., {tag=mouse.screen.selected_tag})
                      end,
                      history_path = awful.util.get_cache_dir() .. "/history"
                  }
              end
```

I also noticed that `awful.widget.prompt()` has some args like `completion_callback` from `awful.prompt` but not others like `exe_callback`, it's unclear which are deprecated...

**Comment by psychon** [comment]
> my "fix" that is probably unreliable:

I would also add `screen = mouse.screen`. Besides that, this is the most reliable thing that you can get, I think.

It might be possible for `awful.startup_notification` to grab the current screen and tag when a SN-sequence begins, but this is still racy since it might take e.g. firefox a while to initiate its SN sequence. So the only reliable way is to specify SN-rules as done in the previous post.

@Elv13 ?

**Comment by Elv13** [comment]
My unreliable fix used in my config is using `awful.spawn{command, {tag=awful.screen.focused().selected_tag})`

**Comment by Elv13** [comment]
> @Elv13 ?

Ugly "fix" in https://github.com/awesomeWM/awesome/pull/1487/commits/5ff74fe43cdcb3cb0938dc6737165b94ba39a1bc (and the other commits included in that PR). With that commit it gets very reliable on Linux, but at the price of an horrible hack. Closing this as a pseudo duplicate of #1482

**Comment by blueyed** [comment]
@Elv13 
Which PR is https://github.com/awesomeWM/awesome/commit/5ff74fe43cdcb3cb0938dc6737165b94ba39a1bc in?
Do you still consider it to be useful?  (I think so)
AFAIU it is meant to grab the snid for processes that set it, but do not follow the protocol itself?!

**Comment by Elv13** [comment]
> Which PR is 5ff74fe in?

It's currently "parked" in https://github.com/awesomeWM/awesome/pull/1487, but eventually would be in its own PR (as it depends on #1487, but is otherwise its own thing.

> AFAIU it is meant to grab the snid for processes that set it, but do not follow the protocol itself?!

Correct. It's also Linux specific and the patch has a couple minor issues (cache not being managed correctly and so on). Otherwise it does work. It leaves only the corner case where:

1) The protocol isn't implemented by the client
2) The client has elaborate ways to keep a single instance or built-in multi process

So it fixes the feature for 99% of the apps, but not from modern browsers... (the second most important apps after the terminals).

> Do you still consider it to be useful? (I think so)

Very, very useful, but also an ugly hack. @psychon What's your opinion on the concept behind that commit?

**Comment by psychon** [comment]
> @psychon What's your opinion on the concept behind that commit?

I'll quote you:

> an ugly hack

If you really want awesome to go systemd (linux only), I think we should be honest about it and say so. Parsing `environ` (badly) doesn't sound too convincing here to me.
